arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/ConcurrentVar.kt

```diff
@@ -258,6 +300,33 @@ private class DefaultConcurrentVar<A> constructor(initial: State<A>) : Concurren
   override suspend fun read(): A =
     cancellable(::unsafeRead)
 
+  override suspend fun <B> withConcurrentVar(f: suspend (A) -> B): B =
+    bracketCase(
+      acquire = ::take,
+      use = f,
+      release = { a, _ -> put(a) }
+    )
+
+  override suspend fun <B> modify(f: suspend (A) -> Tuple2<A, B>): B {
+    // ugly. Is there a better way?
+    var res: A? = null
+    return bracketCase(
+      acquire = ::take,
+      use = {
+        val (a, b) = f(it)
+        res = a
+        b
+      },
+      release = { a, exit ->
+        when (exit) {
+          is ExitCase.Failure -> put(a)
+          is ExitCase.Cancelled -> put(a)
+          is ExitCase.Completed -> put(res!!)
```

Comment:
`res?.let(::put)` maybe?

Reply:
Should `put(res!!)` ever fail for some reason I'd rather have it crash than not put back the value because that could lead to a very confusing deadlock.